File: kotlin/src/main/kotlin/TennisGame4.kt
class TennisGame4(val server: String, val receiver: String) : TennisGame {

    internal var serverScore = 0
    internal var receiverScore = 0

    override fun wonPoint(playerName: String) {
        if (server == playerName) serverScore += 1 else receiverScore += 1
    }

    override fun getScore(): String {
        val result: TennisResult = Deuce(
            this, GameServer(
                this, GameReceiver(
                    this, AdvantageServer(
                        this, AdvantageReceiver(
                            this, DefaultResult(this)
                        )
                    )
                )
            )
        ).result
        return result.format()
    }

    internal fun receiverHasAdvantage(): Boolean {
        return receiverScore >= 4 && receiverScore - serverScore == 1
    }

    internal fun serverHasAdvantage(): Boolean {
        return serverScore >= 4 && serverScore - receiverScore == 1
    }

    internal fun receiverHasWon(): Boolean {
        return receiverScore >= 4 && receiverScore - serverScore >= 2
    }

    internal fun serverHasWon(): Boolean {
        return serverScore >= 4 && serverScore - receiverScore >= 2
    }

    internal fun isDeuce(): Boolean {
        return serverScore >= 3 && receiverScore >= 3 && serverScore == receiverScore
    }
}


internal class TennisResult(var serverScore: String, var receiverScore: String) {
    fun format(): String {
        if ("" == receiverScore) return serverScore
        return if (serverScore == receiverScore) "$serverScore-All" else serverScore + "-" + receiverScore
    }
}

internal interface ResultProvider {
    val result: TennisResult
}

internal class Deuce(private val game: TennisGame4, private val nextResult: ResultProvider) : ResultProvider {
    override val result: TennisResult
        get() = if (game.isDeuce()) TennisResult("Deuce", "") else nextResult.result

}

internal class GameServer(private val game: TennisGame4, private val nextResult: ResultProvider) : ResultProvider {
    override val result: TennisResult
        get() = if (game.serverHasWon()) TennisResult("Win for " + game.server, "") else nextResult.result

}

internal class GameReceiver(private val game: TennisGame4, private val nextResult: ResultProvider) : ResultProvider {
    override val result: TennisResult
        get() = if (game.receiverHasWon()) TennisResult("Win for " + game.receiver, "") else nextResult.result

}

internal class AdvantageServer(private val game: TennisGame4, private val nextResult: ResultProvider) : ResultProvider {
    override val result: TennisResult
        get() = if (game.serverHasAdvantage()) TennisResult("Advantage " + game.server, "") else nextResult.result

}

internal class AdvantageReceiver(private val game: TennisGame4, private val nextResult: ResultProvider) :
    ResultProvider {
    override val result: TennisResult
        get() = if (game.receiverHasAdvantage()) TennisResult("Advantage " + game.receiver, "") else nextResult.result

}

internal class DefaultResult(private val game: TennisGame4) : ResultProvider {
    override val result: TennisResult
        get() = TennisResult(
            scores[game.serverScore], scores[game.receiverScore]
        )

    internal companion object {
        private val scores = arrayOf("Love", "Fifteen", "Thirty", "Forty")
    }
}

File: c/TennisGame4.c
#include <stdlib.h>
#include <string.h>
#include <stdio.h>
#include <stdbool.h>

#include "TennisGame.h"

struct TennisGame {
    int serverScore;
    int receiverScore;
    const char *server;
    const char *receiver;
    char result[18];
};

struct TennisGame *TennisGame_Create(const char *player1, const char *player2) {
    struct TennisGame *newGame = malloc(sizeof(struct TennisGame));
    newGame->serverScore = 0;
    newGame->receiverScore = 0;
    newGame->server = player1;
    newGame->receiver = player2;
    return newGame;
}

void TennisGame_WonPoint(struct TennisGame *game, const char *playerName) {
    if (strcmp(game->server, playerName) == 0)
        game->serverScore += 1;
    else
        game->receiverScore += 1;
}

bool receiverHasAdvantage(struct TennisGame *game) {
    return game->receiverScore >= 4 && (game->receiverScore - game->serverScore) == 1;
}

bool serverHasAdvantage(struct TennisGame *game) {
    return game->serverScore >= 4 && (game->serverScore - game->receiverScore) == 1;
}

bool receiverHasWon(struct TennisGame *game) {
    return game->receiverScore >= 4 && (game->receiverScore - game->serverScore) >= 2;
}

bool serverHasWon(struct TennisGame *game) {
    return game->serverScore >= 4 && (game->serverScore - game->receiverScore) >= 2;
}

bool isDeuce(struct TennisGame *game) {
    return game->serverScore >= 3 && game->receiverScore >= 3 && (game->serverScore == game->receiverScore);
}

struct TennisResult {
    char serverScore[18];
    char receiverScore[18];
};

void formatTennisResult(struct TennisResult *result, char *score) {
    if (strcmp("", result->receiverScore) == 0)
        sprintf(score, "%s", result->serverScore);
    else if (strcmp(result->serverScore, result->receiverScore) == 0)
        sprintf(score, "%s-All", result->serverScore);
    else
        sprintf(score, "%s-%s", result->serverScore, result->receiverScore);
}

typedef struct ResultProvider *ResultProvider_t;
typedef void (*getResult)(struct ResultProvider *resultProvider, struct TennisResult *result);

struct ResultProvider {
    struct TennisGame *game;
    getResult functionPtr;
    ResultProvider_t nextResult;
};

void deuce_getResult(struct ResultProvider *this, struct TennisResult *result) {
    if (isDeuce(this->game))
    {
        strcpy(result->serverScore, "Deuce");
    }
    else
        this->nextResult->functionPtr(this->nextResult, result);
}

void gameServer_getResult(struct ResultProvider *this, struct TennisResult *result) {
    if (serverHasWon(this->game))
    {
        strcpy(result->serverScore, "Win for ");
        strcat(result->serverScore, this->game->server);
    }
    else
        this->nextResult->functionPtr(this->nextResult, result);
}

void gameReceiver_getResult(struct ResultProvider *this, struct TennisResult *result) {
    if (receiverHasWon(this->game))
    {
        strcpy(result->serverScore, "Win for ");
        strcat(result->serverScore, this->game->receiver);
    }
    else
        this->nextResult->functionPtr(this->nextResult, result);
}

void advantageServer_getResult(struct ResultProvider *this, struct TennisResult *result) {
    if (serverHasAdvantage(this->game))
    {
        strcpy(result->serverScore, "Advantage ");
        strcat(result->serverScore, this->game->server);
    }
    else
        this->nextResult->functionPtr(this->nextResult, result);
}

void advantageReceiver_getResult(struct ResultProvider *this, struct TennisResult *result) {
    if (receiverHasAdvantage(this->game))
    {
        strcpy(result->serverScore, "Advantage ");
        strcat(result->serverScore, this->game->receiver);
    }
    else
        this->nextResult->functionPtr(this->nextResult, result);
}

void defaultResult_getResult(struct ResultProvider *this, struct TennisResult *result) {
    const char *scores[] = {"Love", "Fifteen", "Thirty", "Forty"};
    strcpy(result->serverScore, scores[this->game->serverScore]);
    strcpy(result->receiverScore, scores[this->game->receiverScore]);
}

const char *TennisGame_GetScore(struct TennisGame *game) {
    struct TennisResult result;
    result.serverScore[0] = '\0';
    result.receiverScore[0] = '\0';

    struct ResultProvider defaultResult = {game, defaultResult_getResult, NULL};
    struct ResultProvider advantageReceiver = {game, advantageReceiver_getResult, &defaultResult};
    struct ResultProvider advantageServer = {game, advantageServer_getResult, &advantageReceiver};
    struct ResultProvider gameReceiver = {game, gameReceiver_getResult, &advantageServer};
    struct ResultProvider gameServer = {game, gameServer_getResult, &gameReceiver};
    struct ResultProvider deuce = {game, deuce_getResult, &gameServer};

    deuce.functionPtr(&deuce, &result);
    formatTennisResult(&result, game->result);

    return game->result;
}
